### DISCUSSION TITLE
Connection refused when using socket activation

### DISCUSSION THREAD

**Discussion by thyeestes** [containers/podman#23322] [Q&A]
As suggested in the Thread I openend in the Traefik repository the issue below might be related to Pasta or Podman, hence I open a discussion here as well.

What did you do?
I upgraded my Traefik container to v3.1 to enable socket activation. I removed PublishPort from my container file and created two socket files( one for HTTP and one for HTTPS). When my Traefik container is started I can acces my containers from my LAN, but Uptime Kuma cannot reach the websites proxied by Traefik.
I also tested with a other container running a browser but got the same result. All containers are on the same Podman network.

Turning off socket activation solves the issue.

What did you see instead?
Kuma reports an error: connect ECONNREFUSED 192.168.1.100:443 (where as 192.168.1.100 is the IP address of my host running Podman)

I also notice that there are more processes listening on port 80 and 443, so maybe a connection is made to the wrong listener.

Ports listening without socket activation:
rootlessp 589917 thyestes 11u IPv6 2247907 0t0 TCP *:80 (LISTEN)
rootlessp 589917 thyestes 12u IPv6 2247908 0t0 TCP *:443 (LISTEN)

Ports listening with socket activation:
systemd 775 thyestes 34u IPv4 2564674 0t0 TCP *:80 (LISTEN)
systemd 775 thyestes 35u IPv4 2566204 0t0 TCP *:443 (LISTEN)
conmon 666883 thyestes 3u IPv4 2566204 0t0 TCP *:443 (LISTEN)
conmon 666883 thyestes 4u IPv4 2564674 0t0 TCP *:80 (LISTEN)
traefik 666896 thyestes 3u IPv4 2564674 0t0 TCP *:80 (LISTEN)
traefik 666896 thyestes 5u IPv4 2566204 0t0 TCP *:443 (LISTEN)

What version of Traefik are you using?
Version: 3.1.0
Codename: comte
Go version: go1.22.5
Built: 2024-07-15T14:44:04Z
OS/Arch: linux/amd64

What is your environment & configuration?
[Container]
Image=docker.io/library/traefik:v3.1
ContainerName=traefik
Network=proxy.network
Volume=/storage/traefik/acme.json:/acme.json
Volume=%t/podman/podman.sock:/var/run/docker.sock

Add more configuration information here.
Client: Podman Engine
Version: 5.1.2
API Version: 5.1.2
Go Version: go1.21.12
Built: Thu Jul 11 15:01:37 2024
OS/Arch: linux/amd64

I hope there is someone, who can shine a light on this.

**Comment by Luap99** [comment]
Did you allow the ports in your firewall? Can you connect from localhost, i.e. `curl 127.0.0.1`?

**Reply by thyeestes** [reply]
I can connect to the websites from my desktop and my phone, only not from my other containers.

80/tcp                     ALLOW       Anywhere                   # HTTP for Traefik
443/tcp                    ALLOW       Anywhere                   # HTTPS for Traefik

**Reply by Luap99** [reply]
Ah now I understand why you mean. Do you have a second host ip that you can use?
pasta by default duplicates the primary host ip, see https://github.com/containers/podman/issues/22653

**Reply by thyeestes** [reply]
Just to be sure. My setup works flawlessly, when I don't use socket activation, i.e. I use PublishPort in my container file and disable bot socket file.

My host machine is a VM, so I could add an addition network card, if that is what you mean?

**Reply by Luap99** [reply]
The reason it works with the ports is more of an implementation detail, see https://github.com/containers/podman/discussions/22943#discussioncomment-9795883 for how this looks.
The host and rootless-netns have the same ip with pasta and when ports are published we also add DNAT firewall rules in the rootless-netns to redirect traffic from there in the container (like what rootful podman does on the host netns) That is why you do not notice it in this case.

With socket activation the socket was bound on the host which means the everything trying to connect to that must reach the host first. So no direct container to container communicating would work for that I think.

**Reply by Luap99** [reply]
> I could add an addition network card, if that is what you mean?

You do not need a physical network card. A dummy interface with an ip should work fine.

**Reply by thyeestes** [reply]
This is approaching the limits of my knowledge and ability to respond meaningful.

That said, to me it seems that there is no direct container to container communication. I have my Kuma monitors configured like https://servicename.domainname so my container will do a DNS query, gets the IP address of my host (i.e. 192.168.1.100) and connects to port 443 on my host and not directly to my container.

**Reply by thyeestes** [reply]
For the time being I switched back to slirp4netns, which solves the issue.

**Reply by eriksjolund** [reply]
__--map-guest-addr__ is a new option that was added to Pasta last week.
Requirement: passt-0^20240821.g1d6142f or newer. That version was released 21 August 2024.

I wrote an [example](https://github.com/eriksjolund/podman-networking-docs?tab=readme-ov-file#example-connect-to-hosts-main-network-interface-using-pasta-and-custom-network) that shows that setting
```
pasta_options = ["--map-guest-addr","11.11.11.11"]
```
in the file _containers.conf_ allows containers in a custom network to connect to the host's main network interface by connecting to _11.11.11.11_. The IP address _11.11.11.11_ was chosen arbitrarily.

@thyeestes Maybe that option could be useful for this issue?

One thing to solve though is how the Uptime Kuma will look up the hostnames to resolve to _11.11.11.11_

**Reply by Luap99** [reply]
@eriksjolund The plan is to set this by default for the host.containers.internal entry, https://github.com/containers/podman/issues/19213#issuecomment-2305302730

**Reply by eriksjolund** [reply]
@luap99 good to hear! I'll update my example later.

**Reply by thyeestes** [reply]
@eriksjolund I tried switching to Pasta again and added the option you described above. Result is still that Kuma is not able to connect.

**Reply by eriksjolund** [reply]
@thyeestes The required pasta version is from last week (passt-0^20240821.g1d6142f) . Did you try with it?

**Reply by thyeestes** [reply]
Yep that's the version (passt 2024_08_21.1d6142f)

To be sure. I need to replace 11.11.11.11 with the IP address of my host machine, don't I?

**Reply by eriksjolund** [reply]
No, use _11.11.11.11_ or some other IP address except the IP address of your host machine.
I just tried using the IP address of my host machine, but then it doesn't work.

Another thing: Maybe the file _/etc/hosts_ in the Uptime Kuma container needs to be modified so that
the hostnames that Uptime Kuma monitors get resolved to 11.11.11.11?

**Reply by thyeestes** [reply]
Okay tried that, also updated my host file in the Kuma container, however Kuma stil tries to connect to my real host IP address.

**Comment by fsdrw08** [comment]
same here, I use podman socket activation + traefik, I put a hashicorp vault server behind it, and config vault as a OIDC provider, then deploy another app `hashicorp nomad` in the same podman pasta network, when I try to connect from nomad to vault through traefik port 443 which listen by https socket, nomad shows `443: connect: connection refused`, but when I try to connect to the vault with the same https url from my laptop which in the same LAN, it works

**Reply by eriksjolund** [reply]
@fsdrw08 Did you connect using _host.containers.internal_?

Another possibility is to use `--add-host my-vault.example.com:host-gateway` and connect to _my-vault.example.com_ 

(I'm assuming _my-vault.example.com_ resolves in DNS to your host's main network interface IP address)

References:

https://docs.podman.io/en/latest/markdown/podman-run.1.html#add-host-hostname-hostname-ip

https://blog.podman.io/2024/10/podman-5-3-changes-for-improved-networking-experience-with-pasta/

[Outbound TCP/UDP connections to the host's main network interface (e.g eth0)](https://github.com/eriksjolund/podman-networking-docs?tab=readme-ov-file#outbound-tcpudp-connections-to-the-hosts-main-network-interface-eg-eth0)

**Reply by fsdrw08** [reply]
thanks, it works, btw, I am using `podman kube play` to deploy the app, it seems also works to set `host-gateway` in the hostAliases.ip filed
```yaml
hostAliases:
  - ip: "host-gateway"
    hostnames:
      - "vault.example.com"
```

**Comment by eriksjolund** [comment]
I investigated some more:
I think there are two ways a container on the custom network can access the traefik container, either by connecting to the socket-activated sockets (Alternative 2 below) or by connecting to sockets that traefik has created by itself (Alternative 1 below).

[Alternative 1: create extra socket and use NetworkAlias=](https://github.com/eriksjolund/podman-networking-docs?tab=readme-ov-file#alternative-1-create-extra-socket-and-use-networkalias)

[Alternative 2: use AddHost= and host-gateway](https://github.com/eriksjolund/podman-networking-docs?tab=readme-ov-file#alternative-2-use-addhost-and-host-gateway)

An example of Alternative 1 is [example2](https://github.com/eriksjolund/podman-traefik-socket-activation/tree/main/examples/example2) where traefik in addition to inheriting socket-activated sockets from its parent process also creates its own  listening sockets on the custom network. The traefik container is configured with [`NetworkAlias=` ](https://docs.podman.io/en/latest/markdown/podman-systemd.unit.5.html#networkalias) (one for each proxied backend server). One advantage with Alternative 1 is that the download speed is higher when a container downloads files from the traefik container.

**Reply by thyeestes** [reply]
If you add hostnames in your Traefik quadlet, you have to list them for all your containers don't you? Which is quite cumbersome when you have 20+ containers

**Reply by eriksjolund** [reply]
> If you add hostnames in your Traefik quadlet, you have to list them for all your containers don't you?

yes

> Which is quite cumbersome when you have 20+ containers

yes, that is a disadvantage.
